### ISSUE TITLE
custom plugins not recognized in RC4

### ISSUE THREAD

**Issue by dougmcclure** [elastic/logstash#3227]
I normally drop my custom plugins into lib/logstash/output/ and they are recognized as expected in RC2 and RC3. In RC4 upon startup logstash compliains that it can't find the plugin. This was an RC4 RPM install on top of RC2.

Any changes in RC4 that would cause this?

**Comment by untergeek** [comment]
This is because we've made everything in Logstash 1.5 be ruby-gem based, including Logstash core.  See the "How to create a Logstash [input|output|filter|codec] plugin" links at http://www.elastic.co/guide/en/logstash/current/index.html for instructions on how to make your plugin into a ruby gem.

**Comment by dougmcclure** [comment]
Why would it work in RC2 and RC3?  It doesn't make sense to all of a sudden
break/make change in RC4.

Jordan?

**Comment by untergeek** [comment]
Because in RC5 we switched to having a logstash-core gem, different from previous releases.

**Comment by dougmcclure** [comment]
Here's what I get with debug enabled:

plugin not defined in namespace, checking for plugin file {:type=>"output",
:name=>"scala", :path=>"logstash/outputs/scala", :level=>:debug,
:file=>"logstash/plugin.rb", :line=>"133", :method=>"lookup"}
The error reported is:
  Couldn't find any output plugin named 'scala'. Are you sure this is
correct? Trying to load the scala output plugin resulted in this error: no
such file to load -- logstash/outputs/scala
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc4-java/lib/logstash/plugin.rb:142:in
`lookup'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc4-java/lib/logstash/plugin.rb:140:in
`lookup'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc4-java/lib/logstash/pipeline.rb:291:in
`plugin'
(eval):11894:in `initialize'
org/jruby/RubyKernel.java:1107:in `eval'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc4-java/lib/logstash/pipeline.rb:30:in
`initialize'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc4-java/lib/logstash/agent.rb:109:in
`execute'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc4-java/lib/logstash/runner.rb:87:in
`run'
org/jruby/RubyProc.java:271:in `call'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc4-java/lib/logstash/runner.rb:92:in
`run'
org/jruby/RubyProc.java:271:in `call'
/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.19/lib/stud/task.rb:12:in
`initialize'

**Comment by colinsurprenant** [comment]
@wiibaa I am not sure why it works for you... but @dougmcclure behaviour is expected since when searching for `logstash/outputs/...` the load path will point the the `lib/` directory in logstash-core which is now a gem and does not live in the logstash root dir but likely in `vendor/bundle/...`.

I am guessing @wiibaa you are using a cloned repo and @dougmcclure you are using a package?

**Comment by colinsurprenant** [comment]
@dougmcclure but, to answer your question, yes, this changed between rc3 and rc4. In rc4 we completed the gemification of `logstash-core`.

**Comment by dougmcclure** [comment]
So no backwards compat is possible for local, non-gem based plugins from
now on?  Is there even a local install process where connectivity to
external git or gem repositories wouldn't be required?

On Tue, May 12, 2015 at 11:48 AM, Colin Surprenant <notifications@github.com

**Comment by untergeek** [comment]
@dougmcclure the plugin creation docs [here](http://www.elastic.co/guide/en/logstash/current/index.html) document how to create a gem file, which can then be installed via:
```
bin/plugin install /path/to/pluginname.gem
```
So, yes! You'll be able to install plugins without connectivity, or even a local gem repository.

**Comment by colinsurprenant** [comment]
@dougmcclure that's a good point: here's what's possible now-now for you, then we can definitely think about how to (better) support that use-case:

- you could install your output plugin into the logstash-core gem which should be in `vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.snapshot1-java/lib/logstash/output` - let me know if that works for you.

- you could very easily gemify your gem, it's basically just adding a `.gemspec` and then either you build you gem and install it using `bin/plugin install  /path/to/your/scala/scala-xx.gem` or add a path in `Gemfile` with something like:
```
gem "logstash-output-scala", :path => "/path/to/your/scala/" 
```
and run `bin/plugin install --no-verify`

**Comment by colinsurprenant** [comment]
euh, yeah, «gemify you gem» :P meant gemify your plugin...

**Comment by colinsurprenant** [comment]
@dougmcclure a bit anxious in knowing if we could help you fix your local setup with rc4 and if our proposed local gem-based solutions would be viable alternatives for you?

**Comment by dougmcclure** [comment]
Yes, I was able to drop my custom output plugin into here:

/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc4-java/lib/logstash/outputs

and things worked as they did in RC2/RC3.

We'll certainly be moving to the gem based stuff, I just wasn't prepared
for doing that work yet when I decided to test out RC4!

Thanks for the pointers!

Doug

**Comment by colinsurprenant** [comment]
@dougmcclure ah! perfect. sorry for pulling the rug out from under on that one :P 

We'll revisit this to see if we want to add back easier ad-hoc plugin support. Good catch, thanks for reporting it!

**Comment by wiibaa** [comment]
@colinsurprenant I'm using the rc4 zip package not my github clone but I'm on windows.
The only difference I see is `set RUBYLIB=%LS_HOME%\lib` in setup.bat. could it explain it, can it be applied to linux too ?

**Comment by colinsurprenant** [comment]
@wiibaa oh! yeah, right. I guess that's another option, `RUBYLIB` could be set in the env to where ever you want to add ad-hoc plugins. I think we should remove that from the Windows `.bat` scripts since it's now "useless" and make it consistent with Linux and it will free `RUBYLIB` to be set in the env.  I think that would be an acceptable alternative that we could document.

**Comment by colinsurprenant** [comment]
@dougmcclure did you have a chance to try the RUBYLIB env var?

Otherwise, can we close this and followup in another issue for better documenting the options for custom plugins?

**Comment by colinsurprenant** [comment]
FYI we recently had a discussion about this in https://discuss.elastic.co/t/logstash-new-plugin-system-too-restrictive  and it seems the RUBYLIB workaround works for that user.

**Comment by colinsurprenant** [comment]
we decided to restore the `--pluginpath` option, see #3580

**Comment by colinsurprenant** [comment]
closing this as we added back the `--pluginpath` option which will be released with 1.5.3.
